src/expr/src/expr/expr_array_positions.rs

```diff
@@ -0,0 +1,179 @@
+use risingwave_common::util::iter_util::ZipEqFast;
+use risingwave_pb::expr::expr_node::RexNode;
+use risingwave_pb::expr::ExprNode;
+
+use crate::expr::{build_from_prost as expr_build_from_prost, BoxedExpression, Expression};
+use crate::{bail, ensure, ExprError, Result};
+
+pub struct ArrayPositionsExpression {
+    return_type: DataType,
+    left: BoxedExpression,
+    right: BoxedExpression,
+}
+
+impl std::fmt::Debug for ArrayPositionsExpression {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ArrayPositionsExpression")
+            .field("return_type", &self.return_type)
+            .field("left", &self.left)
+            .field("right", &self.right)
+            .finish()
+    }
+}
+
+impl ArrayPositionsExpression {
+    fn new(return_type: DataType, left: BoxedExpression, right: BoxedExpression) -> Self {
+        Self {
+            return_type,
+            left,
+            right,
+        }
+    }
+
+    /// Returns an array of the subscripts of all occurrences of the second argument in the array
+    /// given as first argument. Note the behavior is slightly different from PG.
+    ///
+    /// Examples:
+    ///
+    /// ```slt
+    /// query T
+    /// select array_positions(array[array[1],array[2],array[3],array[2],null::int[]], array[1]);
+    /// ----
+    /// {1}
+    ///
+    /// query T
+    /// select array_positions(array[array[1],array[2],array[3],array[2],null::int[]], array[2]);
+    /// ----
+    /// {2,4}
+    ///
+    /// query T
+    /// select array_positions(array[array[1],array[2],array[3],array[2],null::int[]], null::int[]);
+    /// ----
+    /// {5}
+    ///
+    /// query T
+    /// select array_positions(array[array[1],array[2],array[3],array[2],null::int[]], array[4]);
+    /// ----
+    /// {}
+    ///
+    /// query T
+    /// select array_positions(null::int[], 1);
+    /// ----
+    /// NULL
+    ///
+    /// query T
+    /// select array_positions(ARRAY[array[1],array[2],array[3],array[2],null::int[]], array[3.14]);
+    /// ----
+    /// {}
+    ///
+    /// query T
+    /// select array_positions(array[1,NULL,NULL,3], NULL::int);
+    /// ----
+    /// {2,3}
+    ///
+    /// statement error
+    /// select array_positions(array[array[1],array[2],array[3],array[2],null::int[]], 1);
+    ///
+    /// statement error
+    /// select array_positions(array[array[1],array[2],array[3],array[2],null::int[]], array[array[3]]);
+    ///
+    /// statement error
+    /// select array_positions(ARRAY[array[1],array[2],array[3],array[2],null::int[]], array[true]);
+    /// ```
+    fn evaluate(left: DatumRef<'_>, right: DatumRef<'_>) -> Datum {
+        match left {
+            Some(ScalarRefImpl::List(left)) => {
+                if left.values_ref().len().to_i64().is_some() {
+                    Some(
+                        ListValue::new(
+                            left.values_ref()
```

Comment:
* nit: it seems we are calling `values_ref()` twice, which is unnecessary
* When the result cannot fit into `i64` (or actually i32 to be pg-compatible), we can return an `Err` rather than null. Although it is very likely such an long array cannot even be constructed. (In PG: `select array(select generate_series(0, 2147483647)) is null;`)
* A length check here does not mean the `unwrap` below would never panic, because we add 1 there ... Again this is only theoretical, but why not make it fully correct if cheap?

Reply:
len = max_idx + 1, so the length check here is correct